rust-runtime/smithy-json/src/serialize.rs

```diff
@@ -0,0 +1,451 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+use crate::escape::escape_string;
+use smithy_types::instant::Format;
+use smithy_types::{Instant, Number};
+
+pub struct JsonObjectWriter<'a> {
+    json: &'a mut String,
+    started: bool,
+}
+
+impl<'a> JsonObjectWriter<'a> {
```

Comment:
not sure how this currently plays out in codegen but I _might_ consider:

```rust
pub struct JsonValueWriter<'a> {
  json: &'a mut String,
}

// all writing of values goes through this one place
impl JsonValueWriter {
   pub fn null(self) {
     self.push_str("null");
   }
   pub fn boolean(&self, value) {
      // ...
   }
}

// then instead of having lots of duplicated methods:
impl JsonObjectWriter<'a> {
  pub fn key(&mut self, key: &str) -> JsonValueWriter {
    // write "key": with the comma tracking etc.
    JsonValueWriter { json: self.json }
  }
}

// then the usage would look like:
let mut obj_1 = JsonObjectWriter::new(&mut output);
obj_1.key("foo").boolean(true);
let mut nested = obj_1.key("nested").object();
nested.key("asdf").string("hello!");
nested.finish();
obj_1.finish();
```

Reply:
I really like this idea. I'll simplify down to this in a follow up PR.